File: algorithms/352.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <set>
using namespace std;

struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};

inline bool operator <(const Interval &a, const Interval &b) {
    return a.start < b.start || a.start == b.start && a.end < b.end;
}

class SummaryRanges {
private:
    set<Interval> interval;
    unordered_map<int, bool> isExist;
    unordered_map<int, Interval> leftBound, rightBound;
public:
    /** Initialize your data structure here. */
    SummaryRanges() {
    }

    void addNum(int val) {
        if (isExist[val]) return;
        isExist[val] = true;
        int left = val, right = val;
        if (leftBound.find(val+1) != leftBound.end()) {
            right = leftBound[val+1].end;
            interval.erase(leftBound[val+1]);
            leftBound.erase(val+1);
            rightBound.erase(right);
        }
        if (rightBound.find(val-1) != rightBound.end()) {
            left = rightBound[val-1].start;
            interval.erase(rightBound[val-1]);
            leftBound.erase(left);
            rightBound.erase(val-1);
        }
        Interval i = Interval(left, right);
        interval.insert(i);
        leftBound[left] = i;
        rightBound[right] = i;
        return;
    }

    vector<Interval> getIntervals() {
        return vector<Interval>(interval.begin(), interval.end());
    }
};

int main() {
    return 0;
}

File: algorithms/python3/352.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class SummaryRanges:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.intervals = set()
        self.isExist = set()
        self.leftBound, self.rightBound = {}, {}

    def addNum(self, val):
        """
        :type val: int
        :rtype: void
        """
        if val in self.isExist: return
        self.isExist.add(val)
        left, right = val, val
        if val+1 in self.leftBound:
            right = self.leftBound[val+1].end
            self.intervals.remove(self.leftBound[val+1])
            del self.leftBound[val+1]
            del self.rightBound[right]
        if val-1 in self.rightBound:
            left = self.rightBound[val-1].start
            self.intervals.remove(self.rightBound[val-1])
            del self.leftBound[left]
            del self.rightBound[val-1]
        ivl = Interval(left, right)
        self.intervals.add(ivl)
        self.leftBound[left] = ivl
        self.rightBound[right] = ivl

    def getIntervals(self):
        """
        :rtype: List[Interval]
        """
        return sorted(self.intervals, key=lambda x:x.start)


# Your SummaryRanges object will be instantiated and called as such:
# obj = SummaryRanges()
# obj.addNum(val)
# param_2 = obj.getIntervals()
